src/pages/VenuePage/VenuePage.tsx

```diff
@@ -201,7 +201,7 @@ const VenuePage = () => {
     return () => {
       window.removeEventListener("beforeunload", leaveRoomBeforeUnload, false);
```

Comment:
As per https://github.com/sparkletown/internal-sparkle-issues/issues/99 using `beforeunload` here may lead to us missing events in a number of cases.

> https://developers.google.com/web/updates/2018/07/page-lifecycle-api#legacy-lifecycle-apis-to-avoid
> `unload`, `beforeunload`
>
> Many developers treat the unload event as a guaranteed callback and use it as an end-of-session signal to save state and send analytics data, but doing this is extremely unreliable

Reply:
I completely agree with you. However, visibilitychange and visibilityState don't seem to have a special state for when the tab is closed. And this is actually what we need to uncount the people.

Reply:
From the same URL as above:

> For this reason, it's always better to rely on the `visibilitychange` event to determine when a session ends, and **consider the hidden state the last reliable time to save app and user data.**

Looking at `visibilitychange` it seems as though it will fire in many cases that we don't necessarily care about (eg. the tab not being the focussed, etc). So as you said, it probably isn't suitable for our purposes here.

- https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilitychange_event
- https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilityState

> In all modern browsers (including IE11), **it's recommended to always use the `pagehide` event to detect possible page unloads (a.k.a the terminated state) rather than the `unload` event.** If you need to support Internet Explorer versions 10 and lower, you should feature detect the pagehide event and only use unload if the browser doesn't support pagehide

Looking at `pagehide`, it sounds like it only fires when navigating from the page. Would be worth testing what it does on closing a tab/similar though.. the wording in the description on the following link may just not be ideal, as the description above definitely sounds like it's what we want:

- https://developer.mozilla.org/en-US/docs/Web/API/Window/pagehide_event

Reply:
Looking closer:

- https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event
  - > The `beforeunload` event is fired when the window, the document and its resources are about to be unloaded. The document is still visible and the event is still cancelable at this point.
  - > Attaching an event handler/listener to window or document's `beforeunload` event prevents browsers from using 
  in-memory page navigation caches, like Firefox's Back-Forward cache or WebKit's Page Cache.
  - > Note: You shouldn't use the `beforeunload` event with `sendBeacon()`. See the `Navigator.sendBeacon()` page for more details and best practices.
    - https://developer.mozilla.org/en-US/docs/Web/API/Navigator/sendBeacon
      - > With the `sendBeacon()` method, the data is transmitted asynchronously when the User Agent has an opportunity to do so, without delaying `unload` or the next navigation.
      - > You can use the `PageLifecycle.js` library to deal with current cross-browser inconsistencies in handling of the `visibilitychange` event and other related behavior; that library normalizes cross-browser differences in event-firing order so that state changes always occur exactly as expected, consistently in all browsers.
        - https://github.com/GoogleChromeLabs/page-lifecycle
      - Not sure if this would be relevant or not.. but it feels like we could use it handle uncounting in a nice way, particularly if we are using a firebase function/similar for it (rather than some frontend functionality)
      - This also reminds me, that at one stage I was pondering how we could potentially use firebase function timers/hooks on the backend to potentially handle things like 'timeouts' or similar that we could then use to uncount people. Need to think a bit more about how this would work if they are technically still 'online', but in a zoom room/similar that leaves their browser window 'unattended' though.
- https://developer.mozilla.org/en-US/docs/Web/API/Window/unload_event
  - > It is fired after: `beforeunload` (cancelable event), `pagehide`
- https://developer.mozilla.org/en-US/docs/Web/API/Window/pagehide_event
- https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API
  - > The Page Visibility API provides events you can watch for to know when a document becomes visible or hidden, as well as features to look at the current visibility state of the page.
    - This definitely seems less relevant to our needs/usecases here.. but could be quite useful for optimising things in future (eg. not having to fetch updates/repaint the screen/etc when the page isn't visible, which could reduce CPU load/data usage/etc when it's not needed.
  - https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilitychange_event
    - https://www.w3.org/TR/page-visibility/#dfn-visibilitychange
  - https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilityState
  
Diving deeper into the specs, in case it surfaces any thoughts on better approaches:

- https://html.spec.whatwg.org/multipage/browsing-the-web.html#the-pagetransitionevent-interface
  - `event.persisted`
    - > For the `pageshow` event, returns `false` if the page is newly being loaded (and the `load` event will fire). Otherwise, returns `true`.
    - > **For the `pagehide` event, returns `false` if the page is going away for the last time.** Otherwise, returns `true`, meaning that (if nothing conspires to make the page unsalvageable) the page might be reused if the user navigates back to this page.
- https://html.spec.whatwg.org/multipage/browsing-the-web.html#unloading-documents
  - (this seems to be practically the same content in a different spec: https://www.w3.org/TR/html52/browsers.html#unloading-documents)
  - https://html.spec.whatwg.org/multipage/browsing-the-web.html#unloading-documents:document-3
    - > 5. If the user agent does not intend to keep document alive in a session history entry (such that it can be reused later on history traversal), set document's `salvageable` state to `false`.
    - > 6. Fire an event named `pagehide` at document's relevant global object, using `PageTransitionEvent`, with the `persisted` attribute initialized to `true` if document's `salvageable` state is `true`, and `false` otherwise, and legacy target override flag set.
    - > 8. Unload event: If document's `salvageable` state is `false`, then fire an event named `unload` at document's relevant global object, with legacy target override flag set.
      - This in particular sounds like we can probably use/react to the `pagehide` event, and check whether `salvageable` is set to false instead of using the legacy/deprecated `unload` event.

Reply:
There's actually a nice flowchart at https://developers.google.com/web/updates/2018/07/page-lifecycle-api#overview_of_page_lifecycle_states_and_events that shows the various states/transitions/events that are fired/when:

![image](https://user-images.githubusercontent.com/753891/103949205-04f0df80-518f-11eb-8a01-f75e8aaaf859.png)

Reply:
Captured all of the above in https://github.com/sparkletown/internal-sparkle-issues/issues/99#issuecomment-756413027 so that we don't lose track of it even if we decide it isn't specifically relevant to this PR